File: Sources/Menu.cpp
#include "../Headers/Menu.h"

#if __has_include(<unistd.h>)
const char clearCommand[] = "clear";
#else
const char clearCommand[] = "cls";
#endif

using namespace std;

Menu::Menu(): selectedOption(0)
{
	
}

Menu::~Menu()
{

}

unsigned int Menu::display(const stringstream &summary)
{
	return display(summary.str());
}

unsigned int Menu::display(const string &summary)
{
	char key = 0;
	do
	{
		//Summary display
		cout << summary << endl << endl;

		//Cursor display
		unsigned short int scanIndex = 0;
		for(MenuChoice choice: choiceList)
		{
			string selectedLeft = "	", selectedRight = "	";
			if(scanIndex == selectedOption)
			{
				selectedLeft = "<--	";
				selectedRight = "	-->";
			}
			cout << selectedLeft << choice.getOptionName() << selectedRight << endl;
			scanIndex++;
		}

		//User selection
		key = _getch();
		switch (key)
		{
		case '\n':
		case 13:
			break;

		case 'w':
		case 'z':
		case 'u':
		case '-':
			selectDecr();
			break;

		case 's':
		case 'd':
		case '+':
			selectIncr();
			break;
		default:
			cout << "Unbinded key pressed"/* << " (" << key << ")"*/ << "." << endl << "Press any key to retry." << endl;
			_getch();
		}
		system(clearCommand);
	}while(key != 13 && key != '\n');

	return selectedOption;
}

bool Menu::leaving() const
{
	for (MenuChoice choice : choiceList)
		if(choice.leaves() && choice.getLabel() == selectedOption)
			return true;
	return false;
}

void Menu::selectIncr()
{
	if(selectedOption < choiceList.size() - 1) selectedOption++;
}

void Menu::selectDecr()
{
	if(selectedOption > 0) selectedOption--;
}

void Menu::addChoice(const string &name)
{
	choiceList.push_back(MenuChoice(choiceList.size(), name, false));
}

void Menu::addExit(const string &name)
{
	choiceList.push_back(MenuChoice(choiceList.size(), name, true));
}

bool Menu::deleteChoice(const unsigned int &labelIn)
{
	typename list<MenuChoice>::iterator it = choiceList.begin();
	while (it != choiceList.end())
		if (it++->getLabel() == labelIn)
		{
			choiceList.erase(it);
			return true;
		}
	return false;
}

File: src/MenuPackage/Menu.java
package MenuPackage;

import java.util.ArrayList;

public class Menu
{
	private ArrayList<MenuChoice> _choiceList;
	private int _selectedOption;
	
	public Menu()
	{
		 _selectedOption = 0;
	}
	
	public int display(String summary)
	{
		char key = 0;
		do
		{
			//Summary display
			System.out.println(summary);

			//Cursor display
			int scanIndex = 0;
			for(MenuChoice choice: _choiceList)
			{
				String selectedLeft = "", selectedRight = "";
				if(scanIndex == _selectedOption)
				{
					selectedLeft = ">";
					selectedRight = "<";
				}
				System.out.println(selectedLeft + choice.getOptionName() + selectedRight);
				scanIndex++;
			}

			//User selection
			//key = _getch();[GETCH]
			switch (key)
			{
			case '\n':
			case 13:
				break;

			case 65:
			case 'w':
			case 'z':
			case 'u':
			case '-':
				selectDecr();
				break;

			case 66:
			case 's':
			case 'd':
			case '+':
				selectIncr();
				break;
			default:
				System.out.println("Unbinded key pressed"/* + " (" << key << ")"*/ + ".\nPress any key to retry.");
				//_getch();[GETCH]
			}
			system(clearCommand.c_str());
		}while(key != 13 && key != '\n');

		return _selectedOption;
	}
	
	public boolean leaving()
	{
		for (MenuChoice choice : _choiceList)
			if(choice.leaves() && choice.getLabel() == _selectedOption && !choice.isLocked())
				return true;
		return false;
	}
	
	public void selectIncr()
	{
		if(_selectedOption < _choiceList.size() - 1) _selectedOption++;
	}
	
	public void selectDecr()
	{
		if(_selectedOption > 0) _selectedOption--;
	}
	
	public void addChoice(String name)
	{
		_choiceList.add(new MenuChoice(_choiceList.size(), name, false));
	}
	
	public void addLocked(String name)
	{
		_choiceList.add(
			new MenuChoice(
				_choiceList.size(),
				"[locked] " + name,
				false
			)
		);
	}
	
	public void addExit(String name)
	{
		_choiceList.add(new MenuChoice(_choiceList.size(), name, true));
	}
	
	public boolean deleteChoice(int label)
	{
		int i = 0;
		while (i != _choiceList.size())
			if (_choiceList.get(i++).getLabel() == label)
			{
				_choiceList.remove(i);
				return true;
			}
		return false;
	}
	
	public void clear()
	{
	    _choiceList.clear();
	}
}